Use affine transforms

albfan opened the pull request:
Relay on affine transforms to translate and rotate,
reapplying it on hover, select effect and nob position

## Summary / How this PR fixes the problem?
Using affine transformations allow us to use the internal state of objects to track its changes and apply to related graphic utilities (like nobs or effects)

## Steps to Test
Reproduce this screencast:

![affine-transform](https://user-images.githubusercontent.com/220968/54243914-9c049600-452a-11e9-806c-75266f96ed21.gif)

## Screenshots 

See screencast

## Known Issues / Things To Do
Stopping translate out of visible canvas is missing for simplicity, it can be reworked now all is solved

## This PR fixes/implements the following **bugs/features**:

Rotate objects, which needs refactoring translate and resize

Philip-Scott commented on src/Lib/Canvas.vala:
```diff
@@ -94,31 +94,33 @@ public class Akira.Lib.Canvas : Goo.Canvas {
+
+            event_x_root = canvas_event_x_root;
+            event_y_root = canvas_event_y_root;
+            convert_to_item_space(selected_item, ref event_x_root, ref event_y_root);
```
nit, but add a space before the `(`

Philip-Scott commented on src/Lib/Canvas.vala:
```diff
@@ -153,70 +153,88 @@ public class Akira.Lib.Canvas : Goo.Canvas {
+        var canvas_event_x = event.x;
+        var canvas_event_y = event.y;
+        var event_x = canvas_event_x;
+        var event_y = canvas_event_y;
```
this block of clode was a bit weird. it may be easier to understand if you all setts come from the same variable like:

```
        var canvas_event_x = event.x;
        var canvas_event_y = event.y;
        var event_x = event.x;
        var event_y = event.y;
```

albfan replied:
it tries to reflect `event_y` comes from `canvas_event_x` to be in sync with rest of variable naming.

I tried to get rid of canvas_space or item_space variables, but something don't work as expected. Transforms affect object so item_space is transformed, sometimes is useful (translate) and others is not (rotate)

Let's fix the other obvious things and will come back to this later. I'd want to get rid of it too.

Philip-Scott commented on src/Lib/Canvas.vala:
```diff
@@ -153,70 +153,88 @@ public class Akira.Lib.Canvas : Goo.Canvas {
+        convert_to_item_space(selected_item, ref event_x, ref event_y);
+        var temp_event_x = canvas_temp_event_x;
+        var temp_event_y = canvas_temp_event_y;
+        convert_to_item_space(selected_item, ref temp_event_x, ref temp_event_y);
```
wait a second, you might be having an extra variable here. canvas_temp_event_x was not changed before, so why make a new variable for these? and isn't `temp_event_x` also equal to `event_x`, so why two variables?

albfan replied:
It is at the end of this function. canvas_temp_event_x is the last position processed. Happens that rotate transform is not absolute, but cumulative, so the angle from canvas_event_x_root will not work. canvas_temp_event_x (and its item_space pair temp_event_x) allows to get the angle rotated from the very previous notify_motion_event.

Philip-Scott commented on src/Lib/Canvas.vala:
```diff
@@ -153,70 +153,88 @@ public class Akira.Lib.Canvas : Goo.Canvas {
             case Nob.NONE: // Moving
-                new_x = fix_x_position ((delta_x + start_x), start_w);
-                new_y = fix_y_position ((delta_y + start_y), start_h);
+                selected_item.translate(delta_x, delta_y);
```
Add spaces before the `(` :)

Philip-Scott commented on src/Lib/Canvas.vala:
```diff
@@ -339,24 +361,29 @@ public class Akira.Lib.Canvas : Goo.Canvas {
-        var real_y = y - (line_width * 2);
-        var width = item.bounds.x2 - item.bounds.x1 + stroke - line_width;
-        var height = item.bounds.y2 - item.bounds.y1 + stroke - line_width;
+        var line_width = get_scale() * 2;
```
the division might be correct as it was before. When zooming in the hover effect line is now thicker than before

albfan commented:
Resizing elements is almost working. I need to control better what happens with width and height. Some nobs works correctly right now (RIGHT_BOTTOM, BOTTOM_CENTER,...)

About canvas and item position variables, I think it should be good enough right now, until totally understand goocanvas spaces and maybe find a simpler way to calculate transformations, just to go forward.

albfan commented:
Resize still needs to deal with current_scale, and respect MIN_SIZE on collapse, but at least the "move effect" is gone.

NOTE: there're lot of new vars to track deltas, fixed sizes and fixed deltas. All are supposed to be removed in final implementation

albfan commented:
This needs to deal better with ignored movements from mouse (noticeable when resize get close to borders) and current_scale, but mostly done

Alecaddd commented:
@albfan This is looking really good and it almost works perfectly!
I fixed a small thing related to the stroke size of the hover effect when zooming.

There are some small issues:
* A partially glitchy resize from all the left and top nobs
* When resizing from left to right, the action doesn't stop and moves the object
* When multiple identical shapes are on the canvas, the hover and select effect can't distinguish them

I think we could merge this and then open dedicated issues/PRs to tackle these things separately. What do you think?
I'm good with merging if you want.

albfan commented:
Ok will do a last improvement, and merge it.

El mié., 24 abr. 2019 6:21, Alessandro <notifications@github.com> escribió:

> *@Alecaddd* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akiraux/Akira/pull/121#pullrequestreview-229893543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABV6KFYMNE6SEPXMDWLNM3PR7N2XANCNFSM4G5QOPSQ>
> .
>

albfan commented:
@Alecaddd Still not perfect but:

- A partially glitchy resize from all the left and top nobs
- When resizing from left to right, the action doesn't stop and moves the object

should be mostly solved. (still need polish, but we can do later)

About:

- When multiple identical shapes are on the canvas, the hover and select effect can't distinguish them

Not really sure what it is, so please can you tweak it and push this branch?

Alecaddd commented:
Amazing work!
Let's merge this and then start improving it with small and dedicated PRs.
You, sir, just created a pretty outstanding canvas!

albfan commented:
The best decision. Let's link objects and properties!
